h/config/__init__.py

```diff
@@ -0,0 +1,135 @@
+
+    # Configuration for Pyramid
+    {'setting': 'secret_key', 'env_var': 'SECRET_KEY', 'type': bytes},
+    {'setting': 'secret_salt', 'env_var': 'SECRET_SALT', 'type': bytes, 'default': DEFAULT_SALT},
+
+    # Configuration for h
+    {'setting': 'csp.enabled', 'env_var': 'CSP_ENABLED', 'type': asbool},
+    {'setting': 'csp.report_uri', 'env_var': 'CSP_REPORT_URI'},
+    {'setting': 'csp.report_only', 'env_var': 'CSP_REPORT_ONLY'},
+    {'setting': 'ga_tracking_id', 'env_var': 'GOOGLE_ANALYTICS_TRACKING_ID'},
+    {'setting': 'ga_client_tracking_id', 'env_var': 'GOOGLE_ANALYTICS_CLIENT_TRACKING_ID'},
+    {'setting': 'h.app_url', 'env_var': 'APP_URL'},
+    {'setting': 'h.authority', 'env_var': 'AUTH_DOMAIN',
+     'deprecated_msg': 'use the AUTHORITY environment variable instead'},
+    {'setting': 'h.authority', 'env_var': 'AUTHORITY'},
+    {'setting': 'h.bouncer_url', 'env_var': 'BOUNCER_URL'},
+
+    {'setting': 'h.client_url', 'env_var': 'CLIENT_URL'},
+
+    # ID for the OAuth authclient that the embedded client should use when
+    # making requests to OAuth endpoints. As a public client, it does not have a
+    # secret.
+    {'setting': 'h.client_oauth_id', 'env_var': 'CLIENT_OAUTH_ID'},
+
+    # The list of origins that the client will respond to cross-origin RPC
+    # requests from.
+    {'setting': 'h.client_rpc_allowed_origins',
+     'env_var': 'CLIENT_RPC_ALLOWED_ORIGINS', 'type': aslist},
+
+    {'setting': 'h.db_session_checks', 'env_var': 'DB_SESSION_CHECKS', 'type': asbool},
+
+    # Environment name, provided by the deployment environment. Please do
+    # *not* toggle functionality based on this value. It is intended as a
+    # label only.
+    {'setting': 'h.env', 'env_var': 'ENV'},
+    # Where should logged-out users visiting the homepage be redirected?
+    {'setting': 'h.homepage_redirect_url', 'env_var': 'HOMEPAGE_REDIRECT_URL'},
+    {'setting': 'h.proxy_auth', 'env_var': 'PROXY_AUTH', 'type': asbool},
+    # Sentry DSNs for frontend code should be of the public kind, lacking the
+    # password component in the DSN URI.
+    {'setting': 'h.sentry_dsn_client', 'env_var': 'SENTRY_DSN_CLIENT'},
+    {'setting': 'h.sentry_dsn_frontend', 'env_var': 'SENTRY_DSN_FRONTEND'},
+    {'setting': 'h.websocket_url', 'env_var': 'WEBSOCKET_URL'},
+
+    # Debug/development settings
+    {'setting': 'debug_query', 'env_var': 'DEBUG_QUERY'},
+]
+
+
+def configure(environ=os.environ, settings={}):
+    if 'MANDRILL_USERNAME' in environ and 'MANDRILL_APIKEY' in environ:
+        SETTINGS.extend([
+            {'setting': 'mail.username', 'env_var': 'MANDRILL_USERNAME'},
+            {'setting': 'mail.password', 'env_var': 'MANDRILL_APIKEY'},
+            {'setting': 'mail.host', 'env_var': 'smtp.mandrillapp.com'},
+            {'setting': 'mail.port', 'env_var': 587},
+            {'setting': 'mail.tls', 'env_var': True},
+        ])
+    for s in SETTINGS:
+        try:
```

Comment:
This try catch definitely needs to get removed if we want to crash on startup when we're missing a required env var. However, I wasn't sure exactly how to maintain the existing behavior _and_ crash on startup as they both raise the same exception. We could make multiple exception classes for the various error types, we could simply log a warning inside the env_setting method and not raise an exception, or we could not maintain the existing behavior and raise an exception when the type casting fails (<-I'm inclined to do this last one.).